gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/FSJobCatalog.java

```diff
@@ -0,0 +1,288 @@
+/*
+ * Copyright (C) 2014-2015 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+package gobblin.runtime.job_catalog;
+
+import java.net.URI;
+import java.util.List;
+import java.util.Properties;
+import java.io.IOException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.conf.Configuration;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+
+import gobblin.util.PathUtils;
+import gobblin.runtime.api.JobSpec;
+import gobblin.runtime.api.MutableJobCatalog;
+import gobblin.runtime.api.JobCatalogListener;
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.runtime.util.FSJobCatalogHelper;
+import gobblin.util.filesystem.PathAlterationObserver;
+import gobblin.util.filesystem.PathAlterationDetector;
+import gobblin.util.filesystem.PathAlterationListenerAdaptor;
+import gobblin.runtime.job_catalog.JobCatalogListenersList.AddJobCallback;
+import gobblin.runtime.job_catalog.JobCatalogListenersList.DeleteJobCallback;
+
+
+/**
+ * The job Catalog for file system to persist the job configuration information.
+ * This implementation has no support for caching.
+ */
+public class FSJobCatalog implements MutableJobCatalog {
+  public final Properties jobConfig;
+
+  /**
+   * Listeners of JobCatalog, Like GobblinInstaceDriver.
+   * Distinguished it from JobSpec Listeners.
+   */
+  private final JobCatalogListenersList listeners;
+  private static final Logger LOGGER = LoggerFactory.getLogger(FSJobCatalog.class);
+  /**
+   * The root configuration directory path.
+   */
+  private final Path jobConfDirPath;
+
+  // A monitor for changes to job conf files for general FS
+  // This embedded monitor is monitoring job configuration files instead of JobSpec Object.
+  public final PathAlterationDetector pathAlterationDetector;
+  private final FileSystem fs;
+
+  public Optional<PathAlterationObserver> _observerOptional;
+
+  /**
+   * SINGLEJOB means load job configuration for single job.
+   * BATCHJOB means load a list of job configuration from a directory.
+   */
+  public enum Action {
+    SINGLEJOB, BATCHJOB
+  }
+
+  /**
+   * Initialize the JobCatalog, fetch all jobs in jobConfDirPath.
+   * @param jobConfig
+   * @throws Exception
+   */
+  public FSJobCatalog(Properties jobConfig)
+      throws Exception {
+    this.jobConfig = jobConfig;
+    this.jobConfDirPath = new Path(jobConfig.getProperty(ConfigurationKeys.JOB_CONFIG_FILE_GENERAL_PATH_KEY));
```

Comment:
Please add a precondition check before this line `Preconditions.checkArgument(jobConfig.containsKey(ConfigurationKeys.JOB_CONFIG_FILE_GENERAL_PATH_KEY))` to avoid an NPE